src/exe-mach-o.c

```diff
@@ -92,9 +95,9 @@ int bxfi_exe_patch_main(bxfi_exe_fn *new_main)
     uintptr_t offset = (uintptr_t) addr - (uintptr_t) base;
     size_t len = align2_up(offset + sizeof (opcodes), PAGE_SIZE);
 
-    mprotect(base, len, PROT_READ | PROT_WRITE | PROT_EXEC);
+    mach_vm_protect(mach_task_self(), (mach_vm_address_t) base, len, FALSE, VM_PROT_READ | VM_PROT_WRITE | VM_PROT_EXECUTE | VM_PROT_COPY);
```

Comment:
I'd be happier if there was a feature check for the existence of mach_vm_protect and fallback to mprotect if it isn't available; in essence, adding a new line [there](https://github.com/Snaipe/BoxFort/blob/master/meson.build#L127) and using #ifdef HAVE_MACH_VM_PROTECT here.

Reply:
Thank you very much for the fast response.

I'm not really sure how fragile this call is. `mach_vm_protect()` seems to be actively used in other software (with the undocumented `VM_PROT_COPY` flag too), but I'm a bit uncertain about this. I'm pretty sure this is just a trick and they will eventually patch this out as well.  For this reason, I was thinking whether we should keep the `mprotect()` call as the primary method, and if that fails, we could try falling back to the mach call (if it's available (`HAVE_MACH_VM_PROTECT`)).

What do you think?

I've tried this patch on my really old Macbook (Mojave x86-64), and on Big Sur and Monterey too, they _currently_ work properly.

Reply:
I think I'm happy with doing both on Mac, like in your current commit. It's annoying that we have to work around this bug, but eh.